### PR TITLE
Fixes #9857: ReportingServiceImpl moves around a lot of data and is fairly slow

### PR DESCRIPTION
**Pull request by ncharles**
https://www.rudder-project.org/redmine/issues/9857

### CODE REVIEW

File: rudder-web/src/main/scala/com/normation/rudder/web/snippet/HomePage.scala [hunk 1]
```diff
@@ -267,6 +267,9 @@ class HomePage extends Loggable {
 
       val globalCompliance = compliance.complianceWithoutPending.round
 
+      val n4 = System.currentTimeMillis
+      TimingDebugLogger.debug(s"Compute compliance for HomePage: ${n4 - n2}ms")
```
Comment (ncharles) [inline rudder-web/src/main/scala/com/normation/rudder/web/snippet/HomePage.scala]: i had to move this log, as all compliance are lazy now

**Comment by ncharles** [comment]
Commit modified

File: rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala [hunk 2]
```diff
@@ -109,14 +109,37 @@ final class NodeStatusReport private (
...
   , val report    : AggregatedStatusReport
 ) extends StatusReport {
-  val compliance = report.compliance
-  val byRules: Map[RuleId, AggregatedStatusReport] = report.reports.groupBy(_.ruleId).mapValues(AggregatedStatusReport(_))
+  lazy val compliance = report.compliance
+  lazy val byRules: Map[RuleId, AggregatedStatusReport] = report.reports.groupBy(_.ruleId).mapValues(AggregatedStatusReport(_))
 }
 
 object NodeStatusReport {
   def apply(nodeId: NodeId, runInfo:  RunAndConfigInfo, statusInfo: RunComplianceInfo, reports: Iterable[RuleNodeStatusReport]) = {
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala]: Do we still use the old apply method ?

**Comment by ncharles** [comment]
Commit modified

File: rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala (hunk 2 above)
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala]: yes, it is still used in ReportDisplayer

File: rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala [hunk 3]
```diff
@@ -947,21 +947,17 @@ object ExecutionBatch extends Loggable {
    */
   private[reports] def checkExpectedComponentWithReports(
       expectedComponent: ComponentExpectedReport
-    , filteredReports  : Seq[Reports]
+    , filteredReports  : Seq[ResultReports]
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala]: by design, it is only ResultReports

**Comment by ncharles** [comment]
PR rebased

**Comment by ncharles** [comment]
OK, merging this PR

File: rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala [hunk 4]
```diff
@@ -109,14 +109,37 @@ final class NodeStatusReport private (
...
+  /*
+   * This method filters an NodeStatusReport by Rules.
+   * The goal is to have cheaper way to truncate data from NodeStatusReport that rebuilding it all from top-down
+   * With this approach, we don't need to to the RuleNodeStatusReport merge, and performance are about 20 times better than
+   * recreating and computing all data
+   * it take a NodeStatusReport, and returns it with only the relevant rules
+   */
+  def filterByRules(nodeStatusReport: NodeStatusReport, ruleIds: Set[RuleId]) : NodeStatusReport = {
+    new NodeStatusReport(
+        nodeStatusReport.forNode
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala]: what is this forNode ?

Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala]: ha, this is up there

Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala]: in 4.1 this is nodeId ...

**Comment by ncharles** [comment]
OK, merging this PR

**Pull request merged**